File: music-android-java/app/src/main/java/com/czb/music/MainActivity.java
package com.czb.music;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import android.content.Intent;
import android.os.Bundle;
import android.provider.SyncStateContract;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.alibaba.android.arouter.launcher.ARouter;
import com.czb.module_base.RoutePath;
import com.czb.module_base.base.BaseActivity;
import com.czb.module_base.base.BaseApplication;
import com.czb.module_base.base.BaseFragment;
import com.czb.module_base.callback.LoadingCallback;
import com.czb.module_base.common.Constants;
import com.czb.module_base.common.service.home.wrap.HomeServiceWrap;
import com.czb.module_base.common.service.moyu.wrap.MoyuServiceWrap;
import com.czb.module_base.common.service.ucenter.wrap.UcenterServiceWrap;
import com.czb.module_base.utils.LogUtils;
import com.czb.module_base.utils.SharedPreferencesUtils;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.kingja.loadsir.callback.Callback;
import com.kingja.loadsir.core.LoadService;
import com.kingja.loadsir.core.LoadSir;
import com.lzx.starrysky.StarrySky;

public class MainActivity extends BaseActivity {
    private FragmentManager mFm;
    private BaseFragment mHomeFragment =null;
    private BaseFragment mMoyuFragment =null;
    private BaseFragment mUserFragment = null;
    private boolean isLogin = false;            //是否登录
    private boolean isNeedtoFragmentUser=false;
    private SharedPreferencesUtils mSharedPreferencesUtils;
    private BaseFragment mFragment;
    private BottomNavigationView mNavigationView;

    @Override
    protected void initPresenter() {
    }

    @Override
    protected void initEvent() {
        initListener();
    }

    @Override
    protected void initView() {
//        mNavigationView = findViewById(R.id.main_navigation_bar);
        mSharedPreferencesUtils = SharedPreferencesUtils.getInstance(BaseApplication.getAppContext());
        mNavigationView = this.findViewById(R.id.main_navigation_bar);
        initFragments();
    }




    private void initFragments() {
        mFm = getSupportFragmentManager();
        if (mHomeFragment == null){
            mHomeFragment = (BaseFragment) HomeServiceWrap.Singletion.INSTANCE.getHolder().getFragment();
        }
//        mRecommendFragment = new RecommendFragment();
        if (mMoyuFragment==null){
            mMoyuFragment = (BaseFragment) MoyuServiceWrap.Singletion.INSTANCE.getHolder().getFragment();
        }
        if (mUserFragment==null){
            mUserFragment = (BaseFragment) UcenterServiceWrap.Singletion.INSTANCE.getHolder().getFragment();
        }
//        mUserFragment = new UserFragment();
        //默认选中推荐，不然会出现空白页
        switchFragment(mHomeFragment);
    }



    private void initListener() {
        mNavigationView.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                isLogin = mSharedPreferencesUtils.contains(SharedPreferencesUtils.USER_TOKEN_COOKIE);
                LogUtils.d("test","isLogin ="+isLogin);
                LogUtils.d("test","test");
                if (item.getItemId()==R.id.home){
                    LogUtils.d(MainActivity.this,"切换到首页");
//                        mRecommendFragment = new RecommendFragment();
                    switchFragment(mHomeFragment);
                    return true;
                }else if (item.getItemId()==R.id.moyu){
                    LogUtils.d(MainActivity.this,"切换到社区");
                    switchFragment(mMoyuFragment);
                    return true;
                } else if (item.getItemId()==R.id.user){
                    LogUtils.d(MainActivity.this,"切换到个人");
                    if (isLogin){
//                        LogUtils.d("test",UcenterServiceWrap.Singletion.INSTANCE.getHolder().getFragment().toString()+"");
                        if (mUserFragment ==null){

                            mUserFragment = (BaseFragment) UcenterServiceWrap.Singletion.INSTANCE.getHolder().getFragment();
                        }
                        LogUtils.d("test","logined");
                        switchFragment(mUserFragment);
                        return true;
                    }else {
                        ARouter.getInstance()
                                .build(RoutePath.Login.PATH_lOGIN)
//                                .withString(RoutePath.PATH,RoutePath.Ucenter.FRAGMENT_UCENTER)
                                .navigation();
                        LogUtils.d("test","to login");
                        isNeedtoFragmentUser=true;
                        String token = mSharedPreferencesUtils.getString(SharedPreferencesUtils.USER_TOKEN_COOKIE, "");
                        LogUtils.d("test","logined");
                        switchFragment(lastOneFragment);
                        return false;
                    }
                }
                return true;
            }
        });

    }

    /**
     * 上一次显示的Fragment
     */
    private BaseFragment lastOneFragment = mHomeFragment;

    private void switchFragment(BaseFragment targetFragment) {
        //如果上一个Fragment和当前要切换的fragment是同一个，那么不需要切换
        if (lastOneFragment == targetFragment){
            return;
        }
        //修改成add和hide的方式来控制Fragment
        FragmentTransaction fragmentTransaction = mFm.beginTransaction();
        //如果没有添加就添加
        if (!targetFragment.isAdded()) {
            fragmentTransaction.add(R.id.main_page_container,targetFragment);
        }else {
            fragmentTransaction.show(targetFragment);
        }
        if (lastOneFragment!=null){
            fragmentTransaction.hide(lastOneFragment);
        }
        fragmentTransaction.commit();
        lastOneFragment = targetFragment;

//        //开启事务
//        FragmentTransaction fragmentTransaction = mFm.beginTransaction();
//        fragmentTransaction.replace(R.id.main_page_container,targetFragment);
//        //提交事务
//        fragmentTransaction.commit();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
//        LogUtils.d("onActivityResult","requestCode ="+requestCode+",resultCode="+resultCode+",data = "+data.toString());
        if (resultCode == Constants.RETURN_TO_HMOE){
            switchFragment(mHomeFragment);
            mNavigationView.setSelectedItemId(mNavigationView.getMenu().getItem(Constants.NAVIGATION_VIEW_MENU_RECOMMEND_ITEM_ID_INDEX).getItemId());
        }else if (resultCode==Constants.RETURN_TO_USER){

        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (isNeedtoFragmentUser) {
            isLogin = mSharedPreferencesUtils.contains(SharedPreferencesUtils.USER_TOKEN_COOKIE);
            if (isLogin) {
                if (mUserFragment == null){
                    mUserFragment = (BaseFragment) UcenterServiceWrap.Singletion.INSTANCE.getHolder().getFragment();
                }
                switchFragment(mUserFragment);
                mNavigationView.setSelectedItemId(mNavigationView.getMenu().getItem(Constants.NAVIGATION_VIEW_MENU_USER_ITEM_ID_INDEX).getItemId());
            }else {
                switchFragment(mHomeFragment);
                mNavigationView.setSelectedItemId(mNavigationView.getMenu().getItem(0).getItemId());
            }
        }
    }

    @Override
    public int getLayoutResId() {
        return R.layout.activity_main;
    }

    /**
     * 返回键的事件
     */
    private long exitTime = 0;
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if( keyCode== KeyEvent.KEYCODE_HOME){
            return true;
        } else if( keyCode== KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN){
            if((System.currentTimeMillis()- exitTime) > 2000){
                Toast.makeText(getApplicationContext(), "再按一次退出程序", Toast.LENGTH_SHORT).show();
                exitTime = System.currentTimeMillis();
            } else {
                finish();
                StarrySky.closeNotification();
                System.exit(0);
            }
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }
}

File: music-android/app/src/main/java/com/chen/music/activity/MainActivity.kt
package com.chen.music.activity

import android.util.SparseArray
import android.view.MenuItem
import com.chen.module_base.mvvm.v.BaseActivity
import com.chen.music.MainViewModel
import com.chen.music.databinding.ActivityMainBinding
import com.chen.module_base.utils.EventBusRegister
import androidx.activity.viewModels
import androidx.fragment.app.Fragment
import com.alibaba.android.arouter.launcher.ARouter
import com.chen.module_base.common.constant.RoutePath
import com.chen.module_base.common.constant.SpKey
import com.chen.module_base.common.event.StringEvent
import com.chen.module_base.common.service.community.wrap.CommunityServiceWrap
import com.chen.music.R
import com.chen.module_base.common.service.home.wrap.HomeServiceWrap
import com.chen.module_base.common.service.home.wrap.HomeServiceWrapJava
import com.chen.module_base.common.service.ucenter.wrap.UcenterServiceWrap
import com.chen.module_base.utils.LogUtils
import com.chen.module_base.utils.SpUtils
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

@EventBusRegister
class MainActivity : BaseActivity<ActivityMainBinding,MainViewModel>(){
    override val mViewModel by viewModels<MainViewModel>()

    private var mLastIndex: Int = -1
    private val mFragmentSparseArray = SparseArray<Fragment>()
    // 当前显示的 fragment
    private var mCurrentFragment: Fragment? = null
    private var mLastFragment: Fragment? = null

    override fun ActivityMainBinding.initView() {
        switchFragment(HOME)
        initBottomNavigation()
    }

    private fun initBottomNavigation() {
        mBinding.navView.setOnNavigationItemSelectedListener { menuItem: MenuItem ->
            when (menuItem.itemId) {
                R.id.menu_home -> {
                    switchFragment(HOME)
                    return@setOnNavigationItemSelectedListener true
                }
                R.id.menu_community -> {
                    switchFragment(COMMUNITY)
                    return@setOnNavigationItemSelectedListener true
                }
                R.id.menu_ucenter -> {
                    val isLogin = SpUtils.getBoolean(SpKey.IS_LOGIN, false)

                    if(isLogin == true){
                        switchFragment(UCENTER)
                        return@setOnNavigationItemSelectedListener true
                    }else{
                        ARouter.getInstance()
                            .build(RoutePath.Login.PAGE_LOGIN)
                            .withString(RoutePath.PATH, RoutePath.Ucenter.FRAGMENT_UCENTER)
                            .navigation()
//                        LogUtils.d("MainActivity","isLogin ==> "+isLogin)
//                        ARouter.getInstance().build(RoutePath.Login.PAGE_LOGIN).navigation()
                        return@setOnNavigationItemSelectedListener false
                    }
                }
                else -> return@setOnNavigationItemSelectedListener false
            }
        }
    }
    private fun switchFragment(index: Int) {
        val fragmentManager = supportFragmentManager
        val transaction = fragmentManager.beginTransaction()
        // 将当前显示的fragment和上一个需要隐藏的fragment分别加上tag, 并获取出来
        // 给fragment添加tag,这样可以通过findFragmentByTag找到存在的fragment，不会出现重复添加
        mCurrentFragment = fragmentManager.findFragmentByTag(index.toString())
        mLastFragment = fragmentManager.findFragmentByTag(mLastIndex.toString())
        // 如果位置不同
        if (index != mLastIndex) {
            if (mLastFragment != null) {
                transaction.hide(mLastFragment!!)
            }
            if (mCurrentFragment == null) {
                mCurrentFragment = getFragment(index)
                transaction.add(R.id.content, mCurrentFragment!!, index.toString())
            } else {
                transaction.show(mCurrentFragment!!)
            }
        }

        // 如果位置相同或者新启动的应用
        if (index == mLastIndex) {
            if (mCurrentFragment == null) {
                mCurrentFragment = getFragment(index)
                transaction.add(R.id.content, mCurrentFragment!!, index.toString())
            }
        }
        transaction.commitAllowingStateLoss()
        mLastIndex = index
    }

    private fun getFragment(index: Int): Fragment {
        var fragment: Fragment? = mFragmentSparseArray.get(index)
        if (fragment == null) {
            when (index) {
                HOME -> fragment = HomeServiceWrap.instance.getFragment()
//                HOME -> fragment = HomeServiceWrapJava.Singletion.INSTANCE.holder.fragment
                COMMUNITY -> fragment = CommunityServiceWrap.instance.getFragment()
                UCENTER -> fragment = UcenterServiceWrap.instance.getFragment()
            }
            mFragmentSparseArray.put(index, fragment)
        }
        return fragment!!
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventSwitchTab(event: StringEvent){
        when (event.event) {
            StringEvent.Event.SWITCH_HOME -> mBinding.navView.selectedItemId = R.id.menu_home
            StringEvent.Event.SWITCH_UCENTER -> mBinding.navView.selectedItemId = R.id.menu_ucenter
        }
    }

    override fun initObserve() {
//        TODO("Not yet implemented")
    }

    override fun initRequestData() {
//        TODO("Not yet implemented")
    }


    companion object{
        const val HOME = 0
        const val COMMUNITY = 1
        const val UCENTER = 2
    }
}